File: ts-pr/55/typescript/code-quality-readability-maintainability/game_engine.ts
class GameEngine {
    updateGame(playerX: number, playerY: number, enemyX: number, enemyY: number, 
               playerHealth: { value: number }, enemyHealth: { value: number }, 
               playerAttacking: boolean, enemyAttacking: boolean): void {
        // {fact rule=code-quality-readability-maintainability@v1.0 defects=1}
        if (playerAttacking && Math.abs(playerX - enemyX) < 50 && Math.abs(playerY - enemyY) < 50) {
            enemyHealth.value -= 10;
            if (enemyHealth.value <= 0) {
                console.log('Enemy defeated!');
                return;
            }
        }
        if (enemyAttacking && Math.abs(playerX - enemyX) < 50 && Math.abs(playerY - enemyY) < 50) {
            playerHealth.value -= 15;
            if (playerHealth.value <= 0) {
                console.log('Player defeated!');
                return;
            }
        }
        if (playerX < 0) playerX = 0;
        if (playerX > 800) playerX = 800;
        if (playerY < 0) playerY = 0;
        if (playerY > 600) playerY = 600;
        if (enemyX < 0) enemyX = 0;
        if (enemyX > 800) enemyX = 800;
        if (enemyY < 0) enemyY = 0;
        if (enemyY > 600) enemyY = 600;
        // {/fact}
    }

    updateGameReadable(playerPos: { x: number, y: number }, enemyPos: { x: number, y: number },
                      playerHealth: { value: number }, enemyHealth: { value: number },
                      playerAttacking: boolean, enemyAttacking: boolean): void {
        // {fact rule=code-quality-readability-maintainability@v1.0 defects=0}
        this.handleCombat(playerPos, enemyPos, playerHealth, enemyHealth, playerAttacking, enemyAttacking);
        this.constrainPositions(playerPos, enemyPos);
        // {/fact}
    }

    private handleCombat(playerPos: { x: number, y: number }, enemyPos: { x: number, y: number },
                        playerHealth: { value: number }, enemyHealth: { value: number },
                        playerAttacking: boolean, enemyAttacking: boolean): void {
        const inRange = this.isInAttackRange(playerPos, enemyPos);
        
        if (playerAttacking && inRange) {
            enemyHealth.value -= 10;
            if (enemyHealth.value <= 0) {
                console.log('Enemy defeated!');
            }
        }
        
        if (enemyAttacking && inRange) {
            playerHealth.value -= 15;
            if (playerHealth.value <= 0) {
                console.log('Player defeated!');
            }
        }
    }

    private isInAttackRange(pos1: { x: number, y: number }, pos2: { x: number, y: number }): boolean {
        const attackRange = 50;
        return Math.abs(pos1.x - pos2.x) < attackRange && Math.abs(pos1.y - pos2.y) < attackRange;
    }

    private constrainPositions(playerPos: { x: number, y: number }, enemyPos: { x: number, y: number }): void {
        const screenWidth = 800;
        const screenHeight = 600;
        
        playerPos.x = Math.max(0, Math.min(playerPos.x, screenWidth));
        playerPos.y = Math.max(0, Math.min(playerPos.y, screenHeight));
        enemyPos.x = Math.max(0, Math.min(enemyPos.x, screenWidth));
        enemyPos.y = Math.max(0, Math.min(enemyPos.y, screenHeight));
    }
}

const engine = new GameEngine();
const playerHealth = { value: 100 };
const enemyHealth = { value: 50 };
engine.updateGame(100, 100, 120, 120, playerHealth, enemyHealth, true, false);

File: cpp-pr/46/cpp/code-quality-readability-maintainability/game_engine.cpp
#include <iostream>
#include <string>

class GameEngine {
public:
    void updateGame(int playerX, int playerY, int enemyX, int enemyY, int playerHealth, int enemyHealth, bool playerAttacking, bool enemyAttacking) {
        // {fact rule=code-quality-readability-maintainability@v1.0 defects=1}
        if (playerAttacking && abs(playerX - enemyX) < 50 && abs(playerY - enemyY) < 50) {
            enemyHealth -= 10;
            if (enemyHealth <= 0) {
                std::cout << "Enemy defeated!" << std::endl;
                return;
            }
        }
        if (enemyAttacking && abs(playerX - enemyX) < 50 && abs(playerY - enemyY) < 50) {
            playerHealth -= 15;
            if (playerHealth <= 0) {
                std::cout << "Player defeated!" << std::endl;
                return;
            }
        }
        if (playerX < 0) playerX = 0;
        if (playerX > 800) playerX = 800;
        if (playerY < 0) playerY = 0;
        if (playerY > 600) playerY = 600;
        if (enemyX < 0) enemyX = 0;
        if (enemyX > 800) enemyX = 800;
        if (enemyY < 0) enemyY = 0;
        if (enemyY > 600) enemyY = 600;
        // {/fact}
    }
    
    void updateGameReadable(int playerX, int playerY, int enemyX, int enemyY, int playerHealth, int enemyHealth, bool playerAttacking, bool enemyAttacking) {
        // {fact rule=code-quality-readability-maintainability@v1.0 defects=0}
        handleCombat(playerX, playerY, enemyX, enemyY, playerHealth, enemyHealth, playerAttacking, enemyAttacking);
        constrainPositions(playerX, playerY, enemyX, enemyY);
        // {/fact}
    }
    
private:
    void handleCombat(int playerX, int playerY, int enemyX, int enemyY, int& playerHealth, int& enemyHealth, bool playerAttacking, bool enemyAttacking) {
        bool inRange = isInAttackRange(playerX, playerY, enemyX, enemyY);
        
        if (playerAttacking && inRange) {
            enemyHealth -= 10;
            if (enemyHealth <= 0) {
                std::cout << "Enemy defeated!" << std::endl;
            }
        }
        
        if (enemyAttacking && inRange) {
            playerHealth -= 15;
            if (playerHealth <= 0) {
                std::cout << "Player defeated!" << std::endl;
            }
        }
    }
    
    bool isInAttackRange(int x1, int y1, int x2, int y2) {
        const int ATTACK_RANGE = 50;
        return abs(x1 - x2) < ATTACK_RANGE && abs(y1 - y2) < ATTACK_RANGE;
    }
    
    void constrainPositions(int& playerX, int& playerY, int& enemyX, int& enemyY) {
        const int SCREEN_WIDTH = 800;
        const int SCREEN_HEIGHT = 600;
        
        playerX = std::max(0, std::min(playerX, SCREEN_WIDTH));
        playerY = std::max(0, std::min(playerY, SCREEN_HEIGHT));
        enemyX = std::max(0, std::min(enemyX, SCREEN_WIDTH));
        enemyY = std::max(0, std::min(enemyY, SCREEN_HEIGHT));
    }
};

int main() {
    GameEngine engine;
    engine.updateGame(100, 100, 120, 120, 100, 50, true, false);
    return 0;
}